README.md
## Overview
Сервис по получению и работе с казначейскими счетами, открытыми в Федеральном казначействе
https://roskazna.gov.ru/dokumenty/sistema-kaznacheyskikh-platezhey/kaznacheyskie-scheta/

## Building
...

## Troubleshooting
...

## Release Notes
Can be found in [RELEASE_NOTES](RELEASE_NOTES.md).

## Authors
* Arsen Girinov - [arsengir](https://github.com/arsengir)

## Acknowledgments
...

## Contributing
Please, follow [Contributing](CONTRIBUTING.md) page.

## Code of Conduct
Please, follow [Code of Conduct](CODE_OF_CONDUCT.md) page.

## License
This project is Apache License 2.0 - see the [LICENSE](LICENSE) file for details


RELEASE_NOTES.md
## 1.0-SNAPSHOT
*   Загрузка казначейских счетов с сайта Федерального казначейства


src/test/java/com/github/arsengir/treasuryaccounts/DownloadFileFromURLTest.java
package com.github.arsengir.treasuryaccounts;

import org.junit.jupiter.api.Test;
import java.io.IOException;

import static org.junit.jupiter.api.Assertions.*;

class DownloadFileFromURLTest {

    private final DownloadFileFromURL downloadFileFromURL = new DownloadFileFromURL();

    @Test
    public void getUrlData() {
        String url = null;
        try {
            url = downloadFileFromURL.getUrlData();
        } catch (IOException e) {
            e.printStackTrace();
        }
        System.out.println(url);
        assertNotNull(url);
    }

    @Test
    public void downloadUsingStream() {
        try {
            downloadFileFromURL.downloadUsingStream(downloadFileFromURL.getUrlData());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

src/test/java/com/github/arsengir/treasuryaccounts/TreasuryAccountsUtilTest.java
package com.github.arsengir.treasuryaccounts;

import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.util.LinkedHashSet;

import static org.junit.jupiter.api.Assertions.*;

class TreasuryAccountsUtilTest {

    @Test
    public void getUrlData() {
        String url = null;
        try {
            url = TreasuryAccountsUtil.getUrlData();
        } catch (IOException e) {
            e.printStackTrace();
        }
        System.out.println(url);
        assertNotNull(url);
    }

    @Test
    public void getSetTreasuryAccountsFromLink() {
        LinkedHashSet<TreasuryAccount> treasuryAccountsSet  = new LinkedHashSet<>();
        try {
            treasuryAccountsSet = TreasuryAccountsUtil.getSetTreasuryAccountsFromLink(TreasuryAccountsUtil.getUrlData());
        } catch (IOException e) {
            e.printStackTrace();
        }
        System.out.println(treasuryAccountsSet.size());
        assertFalse(treasuryAccountsSet.isEmpty());
    }

    @Test
    public void saveToFile() {
        TreasuryAccountsUtil.init();
        String fileName = "C:\\Users\\Arsen\\Desktop\\1.txt";
        try {
            TreasuryAccountsUtil.saveToFile(fileName);
        } catch (IOException e) {
            e.printStackTrace();
        }

        assertTrue(new File(fileName).exists());
    }

}

src/main/java/com/github/arsengir/treasuryaccounts/TreasuryAccount.java
package com.github.arsengir.treasuryaccounts;

import java.nio.charset.StandardCharsets;

public class TreasuryAccount {
    // Номер казначейского счета
    private final String TA;
    // Номер единого казначейского счета
    private final String TSA;
    // Банковский идентификационный код территориального органа Федерального казначейства
    // в платежной системе Банка России
    private final String TBIC;
    //Дата открытия казначейского счета
    private final String TAOpenDate;
    //Наименование владельца казначейского счета
    private final String TAHolder;
    //Вариант открытия лицевого счета (LS1 - лицевые счета открыты в территориальном органе Федерального
    //казначейства, LS2 - лицевые счета открыты в финансовом органе или органе управления внебюджетным фондом)
    private final String LSScheme;

    public TreasuryAccount(String ta, String tsa, String tbic, String taOpenDate, String taHolder, String lsScheme) {
        TA = ta;
        TSA = tsa;
        TBIC = tbic;
        TAOpenDate = taOpenDate;
        TAHolder = taHolder;
        LSScheme = lsScheme;
    }

    @Override
    public String toString() {
        return "TreasuryAccount{" +
                "TA='" + TA + '\'' +
                ", TSA='" + TSA + '\'' +
                ", TBIC='" + TBIC + '\'' +
                ", TAOpenDate='" + TAOpenDate + '\'' +
                ", TAHolder='" + TAHolder + '\'' +
                ", LSScheme='" + LSScheme + '\'' +
                '}';
    }

    public byte[] getBytesForSaveToFile() {
        String line = TA + "," + TSA + "," + TBIC + "," + TAOpenDate + "," + TAHolder + "," + LSScheme + "\n";
        return line.getBytes(StandardCharsets.UTF_8);
    }
}


src/main/java/com/github/arsengir/treasuryaccounts/TreasuryAccounts.java
package com.github.arsengir.treasuryaccounts;

public class TreasuryAccounts {
    // Номер казначейского счета
    private final String TA;
    // Номер единого казначейского счета
    private final String TSA;
    // Банковский идентификационный код территориального органа Федерального казначейства
    // в платежной системе Банка России
    private final String TBIC;
    //Дата открытия казначейского счета
    private final String TAOpenDate;
    //Наименование владельца казначейского счета
    private final String TAHolder;
    //Вариант открытия лицевого счета (LS1 - лицевые счета открыты в территориальном органе Федерального
    //казначейства, LS2 - лицевые счета открыты в финансовом органе или органе управления внебюджетным фондом)
    private final String LSScheme;

    public TreasuryAccounts(String ta, String tsa, String tbic, String taOpenDate, String taHolder, String lsScheme) {
        TA = ta;
        TSA = tsa;
        TBIC = tbic;
        TAOpenDate = taOpenDate;
        TAHolder = taHolder;
        LSScheme = lsScheme;
    }
}


src/main/java/com/github/arsengir/treasuryaccounts/DownloadFileFromURL.java
package com.github.arsengir.treasuryaccounts;

import java.io.*;
import java.net.URL;
import java.net.URLConnection;
import java.util.HashMap;

public class DownloadFileFromURL {

    public static final String URL_FILE = "https://www.roskazna.gov.ru/opendata/7710568760-TreasuryAccounts/meta.csv";

    public static String getUrlData() throws IOException {
        URLConnection connection = new URL(URL_FILE).openConnection();
        connection.setRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.95 Safari/537.11");
        connection.connect();

        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
        String line;
        while ((line = reader.readLine()) != null) {
            String[] split = line.split(",");
            if ("\"data\"".equals(split[0])) {
                reader.close();
                return split[1].replaceAll("^\"|\"$", "");
            }
        }
        reader.close();
        return null;
    }

    public static void downloadUsingStream(String urlStr) throws IOException {
        HashMap<String, TreasuryAccounts> hashMap = new HashMap<>();

        URLConnection connection = new URL("https://www.roskazna.gov.ru/upload/iblock/88a/data_20201225T0000_structure_20201225T0000.csv").openConnection();
        connection.setRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.95 Safari/537.11");
        connection.connect();

        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
        String line;
        while ((line = reader.readLine()) != null) {
            String[] split = line.split(",");
            hashMap.put(split[0], new TreasuryAccounts(split[0],split[1],split[2],split[3],split[4],null));
        }
        reader.close();
        hashMap.forEach((k, v) -> System.out.println(k));
    }

}


src/main/java/com/github/arsengir/treasuryaccounts/TreasuryAccountsUtil.java
package com.github.arsengir.treasuryaccounts;

import java.io.*;
import java.net.URL;
import java.net.URLConnection;
import java.util.LinkedHashSet;

public class TreasuryAccountsUtil {

    public static final String URL_FILE = "https://www.roskazna.gov.ru/opendata/7710568760-TreasuryAccounts/meta.csv";

    private static LinkedHashSet<TreasuryAccount> treasuryAccountsSet  = new LinkedHashSet<>();

    public static void init() {
        if (treasuryAccountsSet.isEmpty()) {
            try {
                treasuryAccountsSet = getSetTreasuryAccountsFromLink(getUrlData());
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * Получение ссылки на файл со списком Казначейских счетов
     *
     * @return ссылка на файл со счетами
     * @throws IOException
     */
    public static String getUrlData() throws IOException {
        URLConnection connection = new URL(URL_FILE).openConnection();
        connection.setRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.95 Safari/537.11");
        connection.connect();

        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
        String line;
        while ((line = reader.readLine()) != null) {
            String[] split = line.split(",");
            if ("\"data\"".equals(split[0])) {
                reader.close();
                return split[1].replaceAll("^\"|\"$", "").replaceAll("http://", "https://");
            }
        }
        reader.close();
        return null;
    }

    /**
     * Получение списка (TreasuryAccountsSet) счетов по ссылке
     *
     * @param urlStr ссылка на файл со счетами
     * @throws IOException
     */
    public static LinkedHashSet<TreasuryAccount> getSetTreasuryAccountsFromLink(String urlStr) throws IOException {
        LinkedHashSet<TreasuryAccount> treasuryAccountsSet = new LinkedHashSet<>();

        URLConnection connection = new URL(urlStr).openConnection();
        connection.setRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.95 Safari/537.11");
        connection.connect();

        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
        String line;
        line = reader.readLine();
        if ("TA,TSA,TBIC,TAOpenDate,TAHolder,LSScheme".equals(line)) {
            while ((line = reader.readLine()) != null) {
                String[] split = line.split(",");

                String ta = split[0];
                String tsa = split[1];
                String tbic = split[2];
                String taOpenDate = split[3];
                String taHolder = null;
                String lsScheme = null;

                if (split.length > 4) {
                    int endIndex = split.length - 1;
                    if ("LS1".equals(split[endIndex]) || "LS2".equals(split[endIndex])) {
                        lsScheme = split[endIndex];
                        endIndex--;
                    }

                    StringBuilder sb = new StringBuilder();
                    for (int i = 4; i <= endIndex; i++) {
                        sb.append(split[i]);
                        if (i != endIndex) {
                            sb.append(",");
                        }
                    }
                    taHolder = sb.toString();
                }
                treasuryAccountsSet.add(new TreasuryAccount(ta, tsa, tbic, taOpenDate, taHolder, lsScheme));
            }
            reader.close();
        }
        return treasuryAccountsSet;
    }

    /**
     * Сохранение списка в файл на диск
     * @param fileName - имя файла
     */
    public static void saveToFile(String fileName) throws IOException {
        FileOutputStream outputStream = new FileOutputStream(fileName);
        for (TreasuryAccount ta: treasuryAccountsSet) {
            outputStream.write(ta.getBytesForSaveToFile());
        }
        outputStream.close();
    }

}
